### PR TITLE
Generic HTTP sd

### PR DESCRIPTION
**Pull request by roidelapluie**
This is early WIP based on https://docs.google.com/document/d/1tVeuzjpU4-TiYPNWJXKmcyIuZF6A2tUq270RbBT5zho/edit

The goal is to port the JSON file_sd over HTTP transport.

- [x] Documentation
- [x] Tests (incl errors / config)

### CODE REVIEW

**Comment by rmetzler** [comment]
As a user of Prometheus,  this MR looks really useful to me. So thank you for your work already. 👍 

I'm sorry, my understanding of Go is limited, so I wonder if it already allows to specify HTTP Basic Auth. The use case would be to store the config in a git repo which allows  HTTPS access to <branch>/<file> with username and password (token).

**Review by LeviHarrison — CHANGES_REQUESTED** [review]
Looking really good, just a few documentation things. Also, maybe we could allow the endpoint to respond with `X-Prometheus-Meta-...` headers which could be added to the targets and used for relabeling.

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -1218,6 +1222,81 @@ tls_config:
...
+Each target has a meta label `__meta_url` during the
+[relabeling phase](#relabel_config). Its value is set to the
+url from which the target was extracted.
+
+```yaml
+# URL from wich the targets are fetched.
+url: <string>
+
+# Refresh interval to re-read the files.
+[ refresh_interval: <duration> | default = 5m ]
```
Comment (LeviHarrison) [inline docs/configuration/configuration.md:1261]:
```suggestion
[ refresh_interval: <duration> | default = 60s ]
```
https://github.com/prometheus/prometheus/blob/3bc2b6b3632fc9177d1a618c9d2441b6671e3104/discovery/http/http.go#L40

File: docs/configuration/configuration.md [hunk 2]
```diff
@@ -1218,6 +1222,81 @@ tls_config:
...
+  {
+    "targets": [ "<host>", ... ],
+    "labels": {
+      "<labelname>": "<labelvalue>", ...
+    }
+  },
+  ...
+]
+```
+```
```
Comment (LeviHarrison) [inline docs/configuration/configuration.md:1247]: This is messing up the formatting

```suggestion
```

File: docs/configuration/configuration.md [hunk 3]
```diff
@@ -1218,6 +1222,81 @@ tls_config:
...
+
+Each target has a meta label `__meta_url` during the
+[relabeling phase](#relabel_config). Its value is set to the
+url from which the target was extracted.
+
+```yaml
+# URL from wich the targets are fetched.
+url: <string>
+
+# Refresh interval to re-read the files.
```
Comment (LeviHarrison) [inline docs/configuration/configuration.md:1260]:
```suggestion
# Refresh interval to re-query the endpoints.
```

(or something like that)

**Review by LeviHarrison — COMMENTED** [review]
Looking really good, just a few documentation things. Also, maybe we could allow the endpoint to respond with `X-Prometheus-Meta-...` headers which could be added to the targets and used for relabeling.

**Comment by roidelapluie** [comment]
> Looking really good, just a few documentation things. Also, maybe we could allow the endpoint to respond with `X-Prometheus-Meta-...` headers which could be added to the targets and used for relabeling.

This would break the promise that this is file_sd over http. We do not commit to anything at the HTTP level, labels can be added inside the HTTP body in a more fine-grained way. Also, when debugging, looking at headers is more complicated than body.

File: docs/configuration/configuration.md (hunk 3 above)
Comment (roidelapluie) [reply docs/configuration/configuration.md:1260]: Thanks for the review, will fix this!

**Comment by beorn7** [comment]
Will review once I have read the design doc (which has been on my reading list for a while… :disappointed: ).

**Comment by pakita** [comment]
Hi, 
May I ask  where (release #) can I find this development # 8839. I tried to find it including latest 2.27.1 / 2021-05-18 and could not find truck of this development. Is it still in beta? When we can expect it to be released?
Thank you,
Boris Gdalevich

**Comment by beorn7** [comment]
My review capacity for this week is almost exhausted. Since I haven't even read the design doc yet, I won't be able to do a review before mid next week, realistically. I'd be happy about everyone beating me to it. Sorry…

File: discovery/http/http.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	userAgent = fmt.Sprintf("Prometheus/%s", version.Version)
+)
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+}
+
+// SDConfig is the configuration for HTTP based discovery.
+type SDConfig struct {
```
Comment (owen-d) [inline discovery/http/http.go:51]: nit: would love equivalent `json` tags, even if unused for now as this enables downstream consumers.

**Review by owen-d — COMMENTED** [review]
From the Loki side, this looks great. I don't think we have a pressing need for it currently, but I like the option to plug in SD this way.

File: discovery/http/http.go (hunk 4 above)
Comment (roidelapluie) [reply discovery/http/http.go:51]: Prometheus is only configured in yaml. The previous experience shows that adding json without using it is very error prone. I suggest that all downstream simply align with yaml.

**Comment by beorn7** [comment]
Sorry, I cannot find time for this in my budget. However, both the design doc seem to have competent reviewers.

If another Prometheus team member could finalize the review here?

File: discovery/http/http.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if c.URL == "" {
+		return fmt.Errorf("url missing")
```
Comment (juliusv) [inline discovery/http/http.go:79]: I'd say:

```suggestion
		return fmt.Errorf("URL missing")
```

or:

```suggestion
		return fmt.Errorf("field 'url' missing")
```

File: discovery/http/http.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	parsedURL, err := url.Parse(c.URL)
+	if err != nil {
+		return err
+	}
+	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		return fmt.Errorf("url scheme must be http or https")
```
Comment (juliusv) [inline discovery/http/http.go:86]:
```suggestion
		return fmt.Errorf("URL scheme must be 'http' or 'https'")
```

File: discovery/http/http.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	if parsedURL.Host == "" {
+		return fmt.Errorf("host is missing in URL")
+	}
+	return nil
+}
+
+const httpSDURLLabel = model.MetaLabelPrefix + "url"
+
+// Discovery provides service discovery functionality based
+// on endpoints that return target groups in JSON.
```
Comment (juliusv) [inline discovery/http/http.go:97]:
```suggestion
// on HTTP endpoints that return target groups in JSON format.
```

File: discovery/http/http.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+		logger,
+		"http",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+
```
Comment (juliusv) [inline discovery/http/http.go:133]: Nit: extraneous empty line at the beginning of a method?

File: discovery/http/http.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+		io.Copy(ioutil.Discard, resp.Body)
+		resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("server returned HTTP status %s", resp.Status)
+	}
+
+	if resp.Header.Get("Content-Type") != "application/json" {
+		return nil, errors.Errorf("unsupported content type %s", resp.Header.Get("Content-Type"))
```
Comment (juliusv) [inline discovery/http/http.go:156]:
```suggestion
		return nil, errors.Errorf("unsupported content type %q", resp.Header.Get("Content-Type"))
```

...to get quoting.

**Review by juliusv — COMMENTED** [review]
Looks good to me, besides a few nitty comments :)

**Review by juliusv — APPROVED** [review]
:+1: :ship:

**Comment by juliusv** [comment]
Whoopsie, already approved, but the test error messages now need a fixup still (CI failing). After that it's good to go though.

**Comment by juliusv** [comment]
:+1:

**Comment by pakita** [comment]
Should we expected this functionality with v2.28 that plan to be delivered on 2021-06-16?
Thank you,

**Comment by juliusv** [comment]
Yes!

**Comment by Jeyrce** [comment]
It's a very useful PR, I even did the same thing in my project

**Pull request merged**